File: typescript/src/core/modules/pcustomer-management/functions/get-pcustomers.ts
import { AppError, isStandardError } from "../../../error";
import { PCustomerDAO } from "../data-model/dao";
import { initializeInternalContext } from "../../../context/internal-context";

// Import types
import type { TRuntimeContext } from "../../../context/runtime-context";
import type { TInternalContext } from "../../../context/internal-context";
import type { TFindPCustomerParams } from "../data-model/type";

/**
 * Lấy danh sách các khác hàng ở trong cơ sở dữ liệu.
 *
 * @param ctx - runtime context.
 *
 * @returns
 */
export async function getCustomers(ctx: TRuntimeContext) {
  try {
    const query = await ctx.getQuery<{ limit: string; startKey: string }>();

    const pcustomerDao = new PCustomerDAO();
    const internalCtx =
      initializeInternalContext() as TInternalContext<TFindPCustomerParams>;

    internalCtx.params = {
      limit: query.limit,
      staryKey: query.startKey,
    };
    internalCtx.options!.canCatchError = true;

    const result = await pcustomerDao.listPCustomers(internalCtx);

    return result;
  } catch (error) {
    if (isStandardError(error)) return error;

    const err = new AppError("Cannot found list of potential customers");
    err.asHTTPError("InternalServerError");
    return err;
  }
}

File: python/src/core/modules/pcustomer_management/functions/get_pcustomers.py
# Import from core
from core.context.runtime_context import RuntimeContext
from core.context.internal_context.main import initialize_internal_context
from core.error import AppError, is_standard_error
from core.modules.pcustomer_management.data_model.dao import PCustomerDAO


async def get_pcustomers(ctx: RuntimeContext):
    """
    Lấy thông tin của nhiều khách hàng ở trong hệ thống.

    Args:
        ctx (RuntimeContext): Ngữ cảnh runtime chứa thông tin request.

    Returns:
        dict | AppError: Danh sách thông tin khách hàng hoặc lỗi chuẩn hóa.
    """
    try:
        query = ctx.get_query()

        pcustomer_dao = PCustomerDAO()
        internal_ctx = initialize_internal_context()

        internal_ctx.params = {
            "limit": query.get("limit", "10"),
            "start_key": query.get("startKey"),
        }
        internal_ctx.options["can_catch_error"] = True

        result = await pcustomer_dao.list_pcustomers(internal_ctx)
        return result

    except Exception as error:
        if is_standard_error(error):
            return error

        err = AppError("Cannot find list of potential customers.")
        err.as_http_error("InternalServerError")
        return err
